server/src/main/java/org/opensearch/index/RemoteRefreshSegmentPressureService.java

```diff
@@ -0,0 +1,199 @@
+        if (indexShard.indexSettings().isRemoteStoreEnabled() == false) {
+            return;
+        }
+        trackerMap.remove(shardId);
+        logger.trace("Deleted tracker for shardId={}", shardId);
+    }
+
+    /**
+     * Check if remote refresh segments backpressure is enabled. This is backed by a cluster level setting.
+     *
+     * @return true if enabled, else false.
+     */
+    public boolean isSegmentsUploadBackpressureEnabled() {
+        return pressureSettings.isRemoteRefreshSegmentPressureEnabled();
+    }
+
+    public void validateSegmentsUploadLag(ShardId shardId) {
+        RemoteRefreshSegmentPressureTracker pressureTracker = getPressureTracker(shardId);
+        // Check if refresh checkpoint (a.k.a. seq number) lag is 2 or below - this is to handle segment merges that can
+        // increase the bytes to upload almost suddenly.
+        if (pressureTracker.getSeqNoLag() <= 2) {
+            return;
+        }
+
+        // Check if the remote store seq no lag is above the min seq no lag limit
+        validateSeqNoLag(pressureTracker, shardId);
+
+        // Check if the remote store is lagging more than the upload bytes average multiplied by a variance factor
+        validateBytesLag(pressureTracker, shardId);
+
+        // Check if the remote store is lagging more than the upload time average multiplied by a variance factor
+        validateTimeLag(pressureTracker, shardId);
+
+        // Check if consecutive failure limit has been breached
+        validateConsecutiveFailureLimitBreached(pressureTracker, shardId);
+    }
+
+    private void validateSeqNoLag(RemoteRefreshSegmentPressureTracker pressureTracker, ShardId shardId) {
+        // Check if the remote store seq no lag is above the min seq no lag limit
+        if (pressureTracker.getSeqNoLag() > pressureSettings.getMinSeqNoLagLimit()) {
+            pressureTracker.incrementRejectionCount();
+            throw new OpenSearchRejectedExecutionException(
+                String.format(
+                    Locale.ROOT,
+                    "rejected execution on primary shard:%s due to remote segments lagging behind local segments."
+                        + "remote_refresh_seq_no:%s local_refresh_seq_no:%s",
+                    shardId,
+                    pressureTracker.getRemoteRefreshSeqNo(),
+                    pressureTracker.getLocalRefreshSeqNo()
+                )
+            );
+        }
+    }
+
+    private void validateBytesLag(RemoteRefreshSegmentPressureTracker pressureTracker, ShardId shardId) {
+        if (pressureTracker.isUploadBytesAverageReady() == false) {
+            return;
+        }
+        double dynamicBytesLagThreshold = pressureTracker.getUploadBytesAverage() * pressureSettings.getBytesLagVarianceThreshold();
+        long bytesLag = pressureTracker.getBytesLag();
+        if (bytesLag > dynamicBytesLagThreshold) {
+            pressureTracker.incrementRejectionCount();
```

Comment:
[P1] should we also add separate counter for different rejection reasons

Reply:
Had thought about this, planning to add this in later PRs.

Reply:
@linuxpi  incorporated, do check